### PR TITLE
fix(ecr): allow creating repository uri to use tokens like cfn params

### LINKED ISSUE #31860: lambda.DockerImageCode.fromEcr: imageTagOrDigest field cannot recognize digests supplied as CfnParameter
### Describe the bug

Since imageTagOrDigest supplied in the props for lambda.DockerImageCode.fromEcr can contain either an image tag or digest, if the value isn't explicitly known (in my example it's a CfnParameter) it defaults to it being a tag. There appears to be no workaround for this at the moment

### Regression Issue

- [ ] Select this option if this issue appears to be a regression.

### Last Known Working CDK Version

_No response_

### Expected Behavior

The following cloudformation code should be produced
```
"Code": {
     "ImageUri": {
      "Fn::Join": [
       "",
       [
        {
         "Fn::Select": [
          4,
          {
           "Fn::Split": [
            ":",
            {
             "Ref": "EcrRepoArn"
            }
           ]
          }
         ]
        },
        ".dkr.ecr.",
        {
         "Fn::Select": [
          3,
          {
           "Fn::Split": [
            ":",
            {
             "Ref": "EcrRepoArn"
            }
           ]
          }
         ]
        },
        ".",
        {
         "Ref": "AWS::URLSuffix"
        },
        "/",
        {
         "Ref": "EcrRepoName"
        },
        "@",
        {
         "Ref": "ImageDigest"
        }
       ]
      ]
     }
    },
```

### Current Behavior

The following code is produced. Notice the `:` splitting the image tag
```
"Code": {
     "ImageUri": {
      "Fn::Join": [
       "",
       [
        {
         "Fn::Select": [
          4,
          {
           "Fn::Split": [
            ":",
            {
             "Ref": "EcrRepoArn"
            }
           ]
          }
         ]
        },
        ".dkr.ecr.",
        {
         "Fn::Select": [
          3,
          {
           "Fn::Split": [
            ":",
            {
             "Ref": "EcrRepoArn"
            }
           ]
          }
         ]
        },
        ".",
        {
         "Ref": "AWS::URLSuffix"
        },
        "/",
        {
         "Ref": "EcrRepoName"
        },
        ":",
        {
         "Ref": "ImageDigest"
        }
       ]
      ]
     }
    },
```

### Reproduction Steps

The following code produces an ECR URI with a `:` separating the image digest from the URL instead of an `@`. This causes a deployment failure in cloudformation due to a validation error on the lambda side
```
const imageDigest = new cdk.CfnParameter(this, 'ImageDigest', {
      type: 'String',
      description: 'The image digest',
      default: 'sha256:...',
    });

    const lambdaFunction = new lambda.DockerImageFunction(this, `${lambdaPrefix}`, {
      code: lambda.DockerImageCode.fromEcr(
        ecr.Repository.fromRepositoryAttributes(this, 'RepoCrossAccount', {
          repositoryArn: ecrRepoArn.valueAsString,
          repositoryName: ecrRepoName.valueAsString,
        }), {
          tagOrDigest:  imageDigest.valueAsString,
        }
      ),
    });
```

### Possible Solution

Partition the imageTagOrDigest field into two separate fields or add a type tag 

### Additional Information/Context

_No response_

### CDK CLI Version

2.162.1

### Framework Version

_No response_

### Node.js Version

v22.4.1

### OS

MacOS 14.6.1

### Language

TypeScript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by moelasmar**
### Issue # (if applicable)

Closes #31860.

### Reason for this change
Currently customers can pass one property `tagOrDigest` and if the customers pass a CFN parameter, CDK could not know if it is a tag or digest, and so the generated URI is not correct.

Now the same parameter can supports Tokens, and it will generate a CFN condition to check if the value of this token is digest or tag, and then update the uri based on the condition output.

### Description of changes

Check if the input is a Token, and so instead of determining if its value is a tag or digest in synth time, we create a CFN condition to do this check in CFN, and then determine how to build the repo uri.

### Description of how you validated changes

Added unit test cases, and integration test cases with assertions.

### Checklist
- [X] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-ecr/lib/repository.ts [hunk 1]
```diff
@@ -224,10 +231,31 @@ export abstract class RepositoryBase extends Resource implements IRepository {
    * @param tagOrDigest Optional image tag or digest (digests must start with `sha256:`)
    */
   public repositoryUriForTagOrDigest(tagOrDigest?: string): string {
-    if (tagOrDigest?.startsWith('sha256:')) {
-      return this.repositoryUriForDigest(tagOrDigest);
+    if (tagOrDigest && Token.isUnresolved(tagOrDigest)) {
+
+      let isInputDigestCondition;
+
+      // Use the existing condition of the same Token is existed, and if not create a new one
```
Comment (paulhcsun) [inline packages/aws-cdk-lib/aws-ecr/lib/repository.ts:238]:
```suggestion
      // Use the existing condition of the Token if it already exists, and if not create a new one
```

File: packages/aws-cdk-lib/aws-ecr/lib/repository.ts [hunk 2]
```diff
@@ -224,10 +231,31 @@ export abstract class RepositoryBase extends Resource implements IRepository {
...
+      if (tagOrDigest in this.tokenConditions) {
+        isInputDigestCondition = this.tokenConditions[tagOrDigest];
+      } else {
+        this.conditionsNumber += 1;
+        isInputDigestCondition = new CfnCondition(this, `IsInputDigest${this.conditionsNumber}`, {
+          // we split the value of the Token using the delimiter : to check if the first element equals sha256
+          // to check if it is a digest, or it will be an image tag.
```
Comment (paulhcsun) [inline packages/aws-cdk-lib/aws-ecr/lib/repository.ts:245]:
```suggestion
          // we split the value of the Token using the delimiter : to check if the first element equals sha256
          // in order to determine if it is a digest or an image tag.
```

File: packages/aws-cdk-lib/aws-ecr/lib/repository.ts [hunk 3]
```diff
@@ -224,10 +231,31 @@ export abstract class RepositoryBase extends Resource implements IRepository {
...
+          expression: Fn.conditionEquals(Fn.select(0, Fn.split(':', tagOrDigest)), 'sha256'),
+        });
+        this.tokenConditions[tagOrDigest] = isInputDigestCondition;
+      }
+
+      const suffix = Fn.conditionIf(isInputDigestCondition.logicalId, `@${tagOrDigest}`, `:${tagOrDigest}`).toString();
```
Comment (paulhcsun) [inline packages/aws-cdk-lib/aws-ecr/lib/repository.ts:251]: What is this line doing? What's the purpose of `@${tagOrDigest}` and `:${tagOrDigest}` being passed in?

Comment (moelasmar) [reply packages/aws-cdk-lib/aws-ecr/lib/repository.ts:251]: the condition I added checks if the passed Token contains a Tag or Image Digest, if it is a tag the repo uri should be in this format `<ecr repo base uri>@<image digest value>`, if it is a tag, so the format will be `<ecr repo base uri>:<tag value>`

File: packages/@aws-cdk-testing/framework-integ/test/aws-lambda/test/integ.lambda.image.js.snapshot/aws-cdk-lambda-image.template.json [hunk 4]
```diff
@@ -0,0 +1,761 @@
...
+        },
+        ".",
+        {
+         "Ref": "AWS::URLSuffix"
+        },
+        "/",
+        {
+         "Ref": "MyRepoF4F48043"
+        },
+        "@sha256:55d873154f78cd9a13487c9539ff380f396b887f4a815d84fd698ed872c74749"
```
Comment (moelasmar) [inline packages/@aws-cdk-testing/framework-integ/test/aws-lambda/test/integ.lambda.image.js.snapshot/aws-cdk-lambda-image.template.json:412]: @paulhcsun .. this is an example for the image uri in the Image Digest use case

File: packages/@aws-cdk-testing/framework-integ/test/aws-lambda/test/integ.lambda.image.js.snapshot/aws-cdk-lambda-image.template.json [hunk 5]
```diff
@@ -0,0 +1,761 @@
...
+        ".",
+        {
+         "Ref": "AWS::URLSuffix"
+        },
+        "/",
+        {
+         "Ref": "MyRepoF4F48043"
+        },
+        ":testTag"
+       ]
```
Comment (moelasmar) [inline packages/@aws-cdk-testing/framework-integ/test/aws-lambda/test/integ.lambda.image.js.snapshot/aws-cdk-lambda-image.template.json:643]: @paulhcsun .. and this is an example for the image uri in the tag use case

File: packages/aws-cdk-lib/aws-ecr/lib/repository.ts (hunk 3 above)
Comment (moelasmar) [reply packages/aws-cdk-lib/aws-ecr/lib/repository.ts:251]: I tagged you in 2 examples for each use case in the integ testing template

Comment (moelasmar) [reply packages/aws-cdk-lib/aws-ecr/lib/repository.ts:251]: I can update the code as well to make it more clear

**Review by paulhcsun — APPROVED** [review]

**Pull request merged**